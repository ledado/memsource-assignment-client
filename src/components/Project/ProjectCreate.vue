<template>
  <v-row no-gutters justify="center">
    <v-col cols="4">
      <v-row>
        <v-col cols="12" class="text-h4">
          {{ $t('project.form.create.title') }}
        </v-col>
        <v-col cols="12">
          <project-form
            ref="projectForm"
            v-model="projectToCreate"
            :type="ProjectFormTypeEnum.CREATE"
            :loading="loading"
            @submit="createProject"
          ></project-form>
        </v-col>
        <v-col cols="12">
          <v-btn color="error" :to="{name: $data.routerTypes.MAIN}" block>
            {{ $t('global.cancel') }}
          </v-btn>
        </v-col>
        <v-col cols="12">
          <v-btn color="primary" :to="{name: $data.routerTypes.MAIN}" block>
            {{ $t('project.form.back') }}
          </v-btn>
        </v-col>
      </v-row>
    </v-col>
  </v-row>
</template>

<script lang="ts">
import {Component, Vue} from 'vue-property-decorator';
import ProjectService from '@/service/api/ProjectService';
import {ProjectItem, ProjectStatusEnum, ProjectFormTypeEnum} from '@/service/api/types/ProjectTypes';
import ProjectForm from '@/components/Project/ProjectForm.vue';

@Component({
  components: {ProjectForm},
})
export default class ProjectCreate extends Vue {
  ProjectFormTypeEnum = ProjectFormTypeEnum;
  loading = false;

  projectToCreate: ProjectItem = {
    id: 0,
    name: '',
    sourceLanguage: '',
    status: ProjectStatusEnum.NEW,
    targetLanguages: [],
    dateCreated: '',
    dateUpdated: '',
    dateDue: '',
  };

  $refs!: {
    projectForm: any;
  };

  createProject() {
    this.loading = true;
    const date = new Date();
    ProjectService.createProject({
      ...this.projectToCreate,
      dateCreated: date.toISOString(),
      dateUpdated: date.toISOString(),
      dateDue: date.toISOString(),
    })
      .then(() => {
        this.$toasted.success(this.$t('project.form.create.success').toString());
        this.projectToCreate = {
          id: 0,
          name: '',
          sourceLanguage: '',
          status: ProjectStatusEnum.NEW,
          targetLanguages: [],
          dateCreated: '',
          dateUpdated: '',
          dateDue: '',
        };
        this.$refs.projectForm.resetForm();
      })
      .catch(() => {
        this.$toasted.error(this.$t('global.error.request').toString());
      })
      .finally(() => {
        this.loading = false;
      });
  }
}
</script>
<style lang="scss" scoped></style>
