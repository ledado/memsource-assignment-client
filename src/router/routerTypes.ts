export const routerTypes = {
  MAIN: 'main',
  PROJECT_NEW: 'project-new',
  PROJECT_UPDATE: 'project-update',
};
