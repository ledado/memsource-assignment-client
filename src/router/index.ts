import Vue from 'vue';
import VueRouter, {RouteConfig} from 'vue-router';
import {errorBus} from '@/service/events/ErrorBus';
import {layoutTypes} from '@/layouts/layoutTypes';
import {routerTypes} from '@/router/routerTypes';

Vue.use(VueRouter);

const routes: Array<RouteConfig> = [
  {
    path: '/',
    name: routerTypes.MAIN,
    component: () => import('../views/Main.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
  },
  {
    path: '/project/new',
    name: routerTypes.PROJECT_NEW,
    component: () => import('../views/Project/MainProjectCreate.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
  },
  {
    path: '/project/:id',
    name: routerTypes.PROJECT_UPDATE,
    component: () => import('../views/Project/MainProjectUpdate.vue'),
    meta: {
      layout: layoutTypes.DEFAULT,
    },
    props: true,
  },
  {
    path: '*',
    component: () => import('../views/Error/Error404.vue'),
  },
];

const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes,
});

// error state clean up
router.afterEach((to, from) => {
  if (from.name && errorBus.state) {
    errorBus.errorCode = undefined;
    errorBus.state = false;
  }
});

export default router;
