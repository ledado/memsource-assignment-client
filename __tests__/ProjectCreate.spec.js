import {shallowMount} from '@vue/test-utils';
import ProjectCreate from '@/components/Project/ProjectCreate.vue';
import {routerTypes} from '@/router/routerTypes';

describe('ProjectCreate.vue', () => {
  it('render component', () => {
    const wrapper = shallowMount(ProjectCreate, {
      stubs: ['router-link'],
      mocks: {
        $t: (msg) => msg,
      },
      mixins: [
        {
          data: () => {
            return {
              routerTypes,
            };
          },
        },
      ],
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
