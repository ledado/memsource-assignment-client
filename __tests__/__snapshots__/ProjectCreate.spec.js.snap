// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCreate.vue render component 1`] = `
"<v-row-stub tag=\\"div\\" nogutters=\\"true\\" justify=\\"center\\">
  <v-col-stub cols=\\"4\\" tag=\\"div\\">
    <v-row-stub tag=\\"div\\">
      <v-col-stub cols=\\"12\\" tag=\\"div\\" class=\\"text-h4\\">
        project.form.create.title
      </v-col-stub>
      <v-col-stub cols=\\"12\\" tag=\\"div\\">
        <project-form-stub value=\\"[object Object]\\" type=\\"CREATE\\"></project-form-stub>
      </v-col-stub>
      <v-col-stub cols=\\"12\\" tag=\\"div\\">
        <v-btn-stub color=\\"error\\" tag=\\"button\\" activeclass=\\"\\" to=\\"[object Object]\\" block=\\"true\\" type=\\"button\\">
          global.cancel
        </v-btn-stub>
      </v-col-stub>
      <v-col-stub cols=\\"12\\" tag=\\"div\\">
        <v-btn-stub color=\\"primary\\" tag=\\"button\\" activeclass=\\"\\" to=\\"[object Object]\\" block=\\"true\\" type=\\"button\\">
          project.form.back
        </v-btn-stub>
      </v-col-stub>
    </v-row-stub>
  </v-col-stub>
</v-row-stub>"
`;
